### DISCUSSION TITLE
Overriding comms_prompt_pattern of generic driver

### DISCUSSION THREAD

**Discussion by lucasmarcel** [scrapli/nornir_scrapli#100] [Q&A] [answered]
I have off-brand switches that needs a custom pattern for scrapli to detect the prompt.

I tried adding comms_prompt_pattern under connection_options/scrapli/extras in my groups.yaml file, but the generic driver doesn't seem to pick up on it.

Here is the debug output (redacted):

```
DEBUG:scrapli.driver:load core transport requested
DEBUG:scrapli.driver:core transport 'system' loaded successfully
DEBUG:scrapli.transport:created transport 'open_cmd': '['ssh', '<ip>', '-p', '22', '-o', 'ConnectTimeout=15', '-o', 'ServerAliveInterval=20', '-l', 'user', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-F', '/dev/null', '-o', 'KexAlgorithms=+diffie-hellman-group14-sha1,diffie-hellman-group-exchange-sha1,diffie-hellman-group1-sha1', '-o', 'Ciphers=+aes128-cbc,3des-cbc,aes192-cbc,aes256-cbc']'
INFO:scrapli.driver:opening connection to '<ip>' on port '22'
DEBUG:scrapli.transport:opening transport connection to '<ip>' on port '22'
DEBUG:scrapli.transport:transport connection to '<ip>' on port '22' opened successfully
DEBUG:scrapli.channel:attempting in channel ssh authentication
DEBUG:scrapli.channel:read: b"Warning: Permanently added '<ip>' (RSA) to the list of known hosts.\n"
DEBUG:scrapli.channel:read: b"user@<ip>'s password: "
DEBUG:scrapli.channel:write: REDACTED
DEBUG:scrapli.channel:write: '\n'
DEBUG:scrapli.channel:read: b'\n'
DEBUG:scrapli.channel:read: b'\x1b[1;26HEDS-P510A-8POE series\x1b[1;49HV3.8 build 17041115\x1b[2;1H--\x1b[2J\x1b[1;26HEDS-P510A-8POE series\x1b[1;49HV3.8 build 17041115\x1b[2;1H--\x1b[3;1HEDS-P510A-8PoE-2GTXSFP-T# '
CRITICAL:scrapli.transport:transport operation timed out, closing transport
```

The custom SSH ciphers and algorithms are applied but the pattern is not learned and the connection times out.

If I set the platform to 'ios', the IOSXE driver is invoked but again my prompt pattern is not merged with the others:
```
DEBUG:scrapli:Scrapli factory initialized
INFO:scrapli:Driver '<class 'scrapli.driver.core.cisco_iosxe.sync_driver.IOSXEDriver'>' selected from scrapli core drivers
DEBUG:scrapli.driver:load core transport requested
DEBUG:scrapli.driver:core transport 'system' loaded successfully
DEBUG:scrapli.transport:created transport 'open_cmd': '['ssh', '<ip>', '-p', '22', '-o', 'ConnectTimeout=15', '-o', 'ServerAliveInterval=20', '-l', 'user', '-o', 'StrictHostKeyChecking=no', '-o', 'UserKnownHostsFile=/dev/null', '-F', '/dev/null', '-o', 'KexAlgorithms=+diffie-hellman-group14-sha1,diffie-hellman-group-exchange-sha1,diffie-hellman-group1-sha1', '-o', 'Ciphers=+aes128-cbc,3des-cbc,aes192-cbc,aes256-cbc']'
DEBUG:scrapli.driver:generating combined network comms prompt pattern
DEBUG:scrapli.driver:setting 'comms_prompt_pattern' value to '(^((?!tcl)[a-z0-9.\-_@()/:]){1,63}>$)|(^((?!tcl)[a-z0-9.\-_@/:]){1,63}#$)|(^[a-z0-9.\-_@/:]{1,63}\((?!tcl)[a-z0-9.\-@/:\+]{0,32}\)#$)|((^[a-z0-9.\-_@/:]{1,63}\(tcl\)[>#]$)|(^\+>$))'
INFO:scrapli.driver:opening connection to '<ip>' on port '22'
DEBUG:scrapli.transport:opening transport connection to '<ip>' on port '22'
DEBUG:scrapli.transport:transport connection to '<ip>' on port '22' opened successfully
DEBUG:scrapli.channel:attempting in channel ssh authentication
DEBUG:scrapli.channel:read: b"Warning: Permanently added '<ip>' (RSA) to the list of known hosts.\n"
DEBUG:scrapli.channel:read: b"user@<ip>'s password: "
DEBUG:scrapli.channel:write: REDACTED
DEBUG:scrapli.channel:write: '\n'
DEBUG:scrapli.channel:read: b'\n'
DEBUG:scrapli.channel:read: b'\x1b[1;26HEDS-P510A-8POE series\x1b[1;49HV3.8 build 17041115\x1b[2;1H--\x1b[2J\x1b[1;26HEDS-P510A-8POE series\x1b[1;49HV3.8 build 17041115\x1b[2;1H--\x1b[3;1HEDS-P510A-8PoE-2GTXSFP-T# '
CRITICAL:scrapli.transport:transport operation timed out, closing transport
```

Could you let me know where in the inventory I need to include the prompt pattern? Or maybe it needs to be passed as part of the script?

Thanks.

**Comment by carlmontanari** [comment]
Hi @lucasmarcel off the top of my head I believe the `comms_prompt_pattern` is ignored when using a network driver (because it is overridden by the privilege level patterns). So you have three options!

1. You could make a community platform for your device type -- bit of extra work up front but it isn't too hard and I can help w/ some stuff there.
2. You can use `generic` platform type so you get a `GenericDriver` returned for the connection -- this should not override the `comms_prompt_pattern` you provide.
3. You can provide custom privilege levels to include your pattern -- in nornir this may be a bit more as you'd have to probably spawn the connection then go in and "manually" update the privilege levels/re-run `update_privilege_levels` method. I don't ever use nornir so thats the best I have off the top of my head :)

HTH!

**Reply by lucasmarcel** [reply]
Thanks for your reply. Setting the platform to `generic` is the first thing I tried and it gave me the first debug trace I listed.
Making a community platform sounds great but seems way above my head. I'll read the doc and see if I can understand it.

**Reply by carlmontanari** [reply]
The other thing to look at is the ansi escape codes cropping up in the output. I cant recall off hand if they get (or should get) escaped before or after logging. So may need to do a bit of debugging on there to see whats up there. That may be bricking things before you really get a chance to do much. Can you share what the output looks like just logging into the device "normally" looks like too?

The custom platform is pretty simple -- mostly just a dictionary and two simple functions, so if thats interesting don't be too worried about it being tough!

**Reply by lucasmarcel** [reply]
Sorry for the delay. I figured out that the problem was caused by the 'connection_options/scrapli/extras' dictionary not being merged but instead replaced by the last one defined. So only the options defined as part of defaults.yaml were actually present in my case. The comms_promt_pattern was in the groups.yaml file and therefore ignored. I moved it to defaults.yaml and now the prompt is recognized and everything works as expected.

I understand that this is by design but never ran into it before.
Thanks for your help.

**Reply by carlmontanari** [reply]
Nice, thanks for the update @lucasmarcel , glad you got it sorted! Even easier than making a community platform :D

**Comment by lucasmarcel** [comment, marked as answer]
The problem was caused by the 'connection_options/scrapli/extras' dictionary not being merged but instead replaced by the last one defined. So only the options defined as part of defaults.yaml were actually present in my case. The comms_promt_pattern was in the groups.yaml file and therefore ignored. I moved it to defaults.yaml and now the prompt is recognized and everything works as expected.

I understand that this is by design but never ran into it before.
Thanks for your help.
